R/process_source.R

```diff
@@ -1,131 +1,71 @@
-    if (state == "SCANNING") {
-      if (grepl("^\\S+ *<-", line)) {
-        # Lines that start with "foo <-" tell us that the object definition
-        # starts here.
-        state <- "OBJECT_DEFINITION"
-        current_chunk$add(line)
-        current_chunk_name <- extract_object_name(line)
-
-        # Special case: check if it's a one-liner (one-liners will be parseable;
-        # an opening line for multi-line functions/expressions will not.). If
-        # so, transition to SCANNING state.
-        if (is_parseable(line)) {
-          state <- "SCANNING"
-          chunks[[current_chunk_name]] <- current_chunk$get()
-          current_chunk$reset()
-          current_chunk_name <- NULL
-        }
-
-      } else if (grepl("^#", line)) {
-        state <- "LEADING_COMMENT"
-        current_chunk$add(line)
-      }
-
-    } else if (state == "OBJECT_DEFINITION") {
-      current_chunk$add(line)
-      if (grepl("^\\S", line)) {
-        # If the line starts with anything _other_ than whitespace, then we know
-        # we've reached the end of the object definition.
-        state <- "SCANNING"
-        chunks[[current_chunk_name]] <- current_chunk$get()
-        current_chunk$reset()
-        current_chunk_name <- NULL
-      }
-
-    } else if (state == "LEADING_COMMENT") {
-      if (grepl("^\\S+ *<-", line)) {
-        # Lines that start with "foo <-" tell us that the object definition
-        # starts here.
-        state <- "OBJECT_DEFINITION"
-        current_chunk$add(line)
-        current_chunk_name <- extract_object_name(line)
-
-        # Special case: check if it's a one-liner (one-liners will be parseable;
-        # an opening line for multi-line functions/expressions will not.). If
-        # so, transition to SCANNING state.
-        if (is_parseable(line)) {
-          state <- "SCANNING"
-          chunks[[current_chunk_name]] <- current_chunk$get()
-          current_chunk$reset()
-          current_chunk_name <- NULL
-        }
-
-      } else if (grepl("^#", line)) {
-        # Stay in current state.
-        current_chunk$add(line)
-
-      } else {
-        # We've hit the end of a comment block but no definition here. Just
-        # discard the block and go back to scanning mode.
-        state <- "SCANNING"
-        current_chunk$reset()
-      }
-    }
+# Given a file, process the text and return a named list of objects.
+process_source_file_one <- function(file) {
+  parse_data <- utils::getParseData(parse(file = file, keep.source = TRUE))
+  text <- readLines(file)
+
+  assignment_ops <- parse_data[
+    parse_data$token %in% c("LEFT_ASSIGN", "EQ_ASSIGN", "RIGHT_ASSIGN"),
+  ]
+  # Assignments are the parent expressions of assignment operators
+  assignments <- parse_data[parse_data$id %in% assignment_ops$parent, ]
+  # Top-level expressions have a parent attribute of 0
+  object_definitions <- assignments[assignments$parent == 0, ]
+
+  staticexports <- vector("list", nrow(object_definitions))
```

Comment:
```suggestion
  result <- vector("list", nrow(object_definitions))
```

Also, you don't have to change this, but FWIW, preallocation of the list won't provide much performance benefit, if any, when there's only a single reference to the object, as there is in this code. With recent R versions, it will grow the list in place (if possible, which is true most of the time) when there's only one reference to it.

Reply:
I didn't know that! Updated in [26cd7de](https://github.com/wch/staticimports/pull/21/commits/26cd7de5a3966f88f845b38cc6781447eced8326)

Reply:
Here's some example code demonstrating this:
https://gist.github.com/wch/c5642fa8aef65c9c1d3b076dc9e2b813